src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -210,14 +220,32 @@ private void TrackOpenedDocument(uint docCookie, string moniker)
 
         private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
+            // We may or may not be getting this notification from the foreground thread if another workspace
+            // is raising events on a background. Let's send it back to the UI thread since we can't talk
+            // to the RDT in the background thread. Since this is all asynchronous a bit more asynchrony is fine.
+            if (!_foregroundThreadAffinitization.IsForeground())
+            {
+                ScheduleTask(() => Registration_WorkspaceChanged(sender, e));
```

Comment:
@Pilchie The fix comes from checking `_runningDocumentTable.IsCookieValid(docCookie)` since we are given a bad cookie from `_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie)`.  I believe the foreground thread asserts are inconsequential to this fix, though I haven't checked thoroughly.  Devenv should be calling us from a foreground thread, but in such a way that we receive this event in error as the RDT has not opened anything.  I believe this is due to us caching workspace info in the .vs folder as this bug does not reproduce unless a precious .vs folder exists and we have opened documents in the solution.